libdmet/basis_transform/eri_transform.py

```diff
@@ -257,33 +262,39 @@ def get_emb_eri_fast_gdf(cell, mydf, C_ao_lo=None, basis=None, feri=None,
     # treat the possible drop of aux-basis at some kpts.
     naux = get_naoaux(mydf)
 
-    # If C_ao_lo and basis not given, this routine is k2gamma AO transformation
-    if C_ao_lo is None:
-        C_ao_lo = np.zeros((nkpts, nao, nao), dtype=np.complex128)
-        C_ao_lo[:, range(nao), range(nao)] = 1.0 # identity matrix for each k
-
-    # add spin dimension for restricted C_ao_lo
-    if C_ao_lo.ndim == 3:
-        C_ao_lo = C_ao_lo[np.newaxis]
-
     # possible kpts shift
     kscaled = cell.get_scaled_kpts(kpts)
     if kscaled_center is not None:
         kscaled -= kscaled_center
 
-    # basis related
-    if basis is None:
-        basis = np.eye(nkpts * nao).reshape(1, nkpts, nao, nkpts * nao)
-    if basis.shape[0] < C_ao_lo.shape[0]:
-        basis = add_spin_dim(basis, C_ao_lo.shape[0])
-    if C_ao_lo.shape[0] < basis.shape[0]:
-        C_ao_lo = add_spin_dim(C_ao_lo, basis.shape[0])
-
-    if unit_eri: # unit ERI for DMFT
-        C_ao_emb = C_ao_lo / (nkpts**0.75)
+    if C_ao_eo is None:
+        # If C_ao_lo and basis not given, this routine is k2gamma AO transformation
+        if C_ao_lo is None:
+            C_ao_lo = np.zeros((nkpts, nao, nao), dtype=np.complex128)
+            C_ao_lo[:, range(nao), range(nao)] = 1.0 # identity matrix for each k
+
+        # add spin dimension for restricted C_ao_lo
+        if C_ao_lo.ndim == 3:
+            C_ao_lo = C_ao_lo[np.newaxis]
+
+        # basis related
+        if basis is None:
+            basis = np.eye(nkpts * nao).reshape(1, nkpts, nao, nkpts * nao)
+        if basis.shape[0] < C_ao_lo.shape[0]:
+            basis = add_spin_dim(basis, C_ao_lo.shape[0])
+        if C_ao_lo.shape[0] < basis.shape[0]:
+            C_ao_lo = add_spin_dim(C_ao_lo, basis.shape[0])
+
+        if unit_eri: # unit ERI for DMFT
+            C_ao_emb = C_ao_lo / (nkpts**0.75)
+        else:
+            phase = get_phase_R2k(cell, kpts)
+            C_ao_emb = multiply_basis(C_ao_lo, get_basis_k(basis, phase)) / (nkpts**(0.75))
     else:
-        phase = get_phase_R2k(cell, kpts)
-        C_ao_emb = multiply_basis(C_ao_lo, get_basis_k(basis, phase)) / (nkpts**(0.75))
+        assert C_ao_lo is None
+        assert (nkpts, nao) == C_ao_eo.shape[:2]
```

Comment:
I think this PR is overall good. Just one question: Can `C_ao_eo` itself have spin dimension?

Reply:
As far as I know not. :-(

Reply:
I think the following should work,

```
        assert C_ao_lo is None
        C_ao_eo = np.asarray(C_ao_eo)
        if C_ao_eo.ndim == 3:
             C_ao_eo = C_ao_eo[np.newaxis]
        assert (nkpts, nao) == C_ao_eo.shape[1:3]
        C_ao_emb = C_ao_eo / (nkpts**(0.75))
```

Reply:
Yeah, makes sense.
https://github.com/gkclab/libdmet_preview/pull/21/commits/0311d2db969979b1318e3b6272f1c34337064244